### PR TITLE
Sri auth

### PR DESCRIPTION
**Pull request by Manik-Jain**
# Description

This PR has changes for SRI authentication - backend and ReactJs frontend

## Related Issue

https://github.com/eea-oasis/baseline/issues/588
https://github.com/eea-oasis/baseline/issues/589
https://github.com/eea-oasis/baseline/issues/590

## Motivation and Context

In order to only authenticated access, we need to have a mandatory bearer token for each end-point

## How Has This Been Tested

Local tested

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Request to be added as a Code Owner/Maintainer

## Checklist

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.
-[X] I commit to abide by the Responsibilities of Code Owners/Maintainers.

### CODE REVIEW

File: examples/bri-3/src/bri/auth/api/auth.controller.ts
Comment (ognjenkurtic) [inline examples/bri-3/src/bri/auth/api/auth.controller.ts:19]: If someone knows a message and a signature of that message for a specific public key, they can pass the login as that public key. This is because the mnemonic (one time pass)  is not stored during geneation so we do not check if it was issued by our backend during login. Do we really want this?

@Manik-Jain @skosito @Therecanbeonlyone1969 @Kasshern @biscuitdey @Ybittan

File: examples/bri-3/front-end/src/login/Login.js
Comment (skosito) [inline examples/bri-3/front-end/src/login/Login.js:24]: in general its bad practice to store token in local storage but i guess it will do for this

File: examples/bri-3/package.json
Comment (skosito) [inline examples/bri-3/package.json:34]: still reviewing but 10 more dependencies for this seems a lot, why we need passport and nest jwt?

File: examples/bri-3/src/bri/auth/agents/authAccessor.agent.ts
Comment (skosito) [inline examples/bri-3/src/bri/auth/agents/authAccessor.agent.ts:5]: i dont think new service is needed, this looks like it should be part of bpi subject storage agent
authAccessor.findOne doesnt say anything about bpi subject being returned

File: examples/bri-3/src/bri/auth/api/auth.controller.ts
Comment (skosito) [reply examples/bri-3/src/bri/auth/api/auth.controller.ts:19]: yea... we need to store some nonce that is signed and then when authenticating pull the last nonce from db, not send it in request body
also, why using bip39 mnemonic at first place as thing to be signed? lets just store some incrementing nonce or uuid or something like that...

File: examples/bri-3/src/bri/auth/guards/sri-jwt.guard.ts
Comment (skosito) [inline examples/bri-3/src/bri/auth/guards/sri-jwt.guard.ts:13]: what is this doing? please add comment

File: examples/bri-3/src/bri/auth/guards/sri-local.guard.ts
Comment (skosito) [inline examples/bri-3/src/bri/auth/guards/sri-local.guard.ts:5]: why we need both jwt and local auth guards?

File: examples/bri-3/src/bri/auth/services/auth.service.ts
Comment (skosito) [inline examples/bri-3/src/bri/auth/services/auth.service.ts:17]: use bpiSubject terminology instead of user, use publicKey instead of username

why do we need password at all? or verifyUser method?

File: examples/bri-3/src/bri/auth/model/login.ts
Comment (skosito) [inline examples/bri-3/src/bri/auth/model/login.ts:3]: what is Login, is it dto or? if its request body, please use dto convention we use, with not empty validators etc
also message should not be part of this request body, but pulled out of db, check comments above

File: examples/bri-3/src/bri/auth/services/auth.service.ts
Comment (skosito) [inline examples/bri-3/src/bri/auth/services/auth.service.ts:11]: we introduce 4 services that are at least somehow  connected
- auth accessor service
- jwt service
- auth service
- login service

why so many services for simple auth flow?

File: examples/bri-3/src/bri/auth/services/login.service.ts
Comment (skosito) [inline examples/bri-3/src/bri/auth/services/login.service.ts:5]: see my comment about 4 services for auth above...

File: examples/bri-3/src/bri/auth/strategy/jwt.strategy.ts
Comment (skosito) [inline examples/bri-3/src/bri/auth/strategy/jwt.strategy.ts:1]: importing nestjs/passport
passport
passport-jwt
nestjs/jwt

in same project? do we need all those packages?

File: examples/bri-3/src/bri/auth/strategy/local.strategy.ts
Comment (skosito) [inline examples/bri-3/src/bri/auth/strategy/local.strategy.ts:8]: what is the difference between local and jwt strategies? why do we need both?

File: examples/bri-3/src/bri/identity/bpiSubjects/agents/bpiSubjectsStorage.agent.ts
Comment (skosito) [inline examples/bri-3/src/bri/identity/bpiSubjects/agents/bpiSubjectsStorage.agent.ts:30]: please remove console.logs before merging

File: examples/bri-3/src/bri/identity/bpiSubjects/agents/bpiSubjectsStorage.agent.ts
Comment (skosito) [inline examples/bri-3/src/bri/identity/bpiSubjects/agents/bpiSubjectsStorage.agent.ts:48]: no need to store as lower case, only lower case both on places we compare

File: examples/bri-3/src/bri/decorators/public-endpoint.ts
Comment (skosito) [inline examples/bri-3/src/bri/decorators/public-endpoint.ts:1]: what is public endpoint?

**Review by skosito — CHANGES_REQUESTED** [review]

File: examples/bri-3/src/bri/decorators/public-endpoint.ts
Comment (Manik-Jain) [reply examples/bri-3/src/bri/decorators/public-endpoint.ts:1]: in order for login, or any, function to be invoked without any bearer token we need to have a way. This decorator does that job for us.

File: examples/bri-3/src/bri/identity/bpiSubjects/agents/bpiSubjectsStorage.agent.ts
Comment (Manik-Jain) [reply examples/bri-3/src/bri/identity/bpiSubjects/agents/bpiSubjectsStorage.agent.ts:48]: as a general observation, ethers and other library returns us the lowercase MM public key.

File: examples/bri-3/src/bri/auth/strategy/local.strategy.ts
Comment (Manik-Jain) [reply examples/bri-3/src/bri/auth/strategy/local.strategy.ts:8]: js doc was already added to the classes.

File: examples/bri-3/src/bri/auth/strategy/jwt.strategy.ts
Comment (Manik-Jain) [reply examples/bri-3/src/bri/auth/strategy/jwt.strategy.ts:1]: I would suggest reading through this document: https://docs.nestjs.com/security/authentication#authentication

File: examples/bri-3/src/bri/auth/services/auth.service.ts
Comment (Manik-Jain) [reply examples/bri-3/src/bri/auth/services/auth.service.ts:11]: No, they are not at all connected. Each service has its own defined responsibility.

File: examples/bri-3/package.json
Comment (Manik-Jain) [reply examples/bri-3/package.json:34]: kindly skim read here: https://docs.nestjs.com/security/authentication#authentication

File: examples/bri-3/src/bri/auth/agents/authAccessor.agent.ts
Comment (Manik-Jain) [reply examples/bri-3/src/bri/auth/agents/authAccessor.agent.ts:5]: I wanted to keep auth related code inside `auth` folder, and leave other CRUD operations inside identity. I don't this change is causing any blockers

File: examples/bri-3/src/bri/auth/guards/sri-local.guard.ts
Comment (Manik-Jain) [reply examples/bri-3/src/bri/auth/guards/sri-local.guard.ts:5]: Right, we just need jwt guard. Removing the other one

File: examples/bri-3/src/bri/auth/services/auth.service.ts
Comment (Manik-Jain) [reply examples/bri-3/src/bri/auth/services/auth.service.ts:17]: Removed

File: examples/bri-3/src/bri/auth/model/login.ts
Comment (Manik-Jain) [reply examples/bri-3/src/bri/auth/model/login.ts:3]: changed

File: examples/bri-3/front-end/src/login/Login.js
Comment (Manik-Jain) [reply examples/bri-3/front-end/src/login/Login.js:24]: Well, there is a mixed approach that people/teams use. Anyways, for the current scope, this can be removed

File: examples/bri-3/src/bri/auth/api/auth.controller.ts
Comment (Manik-Jain) [reply examples/bri-3/src/bri/auth/api/auth.controller.ts:19]: @ognjenkurtic Right, I missed on the OTP storage part. Will push in the code soon.
@skosito the incrementing nonce approach works, and I think this is legacy. bip39 mnemonics can act as the random string of words to be signed, and updated in DB  the hash for each new generated mnemonic, similar to a nonce.

Another approach to it can be, we store in the signature instead of the nonce/mnemonic, to identify if a jwt token was linked to the provided signature. This way, we cover both the scenario

**Review by ognjenkurtic — CHANGES_REQUESTED** [review]
Some general comments:

1) Please follow the branch naming standard of the Baseline repository.

2) Please provide more meaningful commit messages. General agreement is to do "This commit does..." approach where your commit message just continues the sentence. Also, commits should be much smaller.

3) Please follow the process we agreed on in the beginning of the project for defining the implementation direction. 

           Create a doc with proposal -> Iterate with the team -> Write down final agreement -> Create separate smaller issues for implementation -> Implement by doing a single PR per issue

This is important for a multitude of reasons, most notably:
   - We document our design decisions in a transparent and consistent way
   - We do most of the design during the initial iterations by the team, avoiding costly rewrites and discussions once PR is created
   - We produce smaller PRs, easier to review thus with higher quality

4) Please follow the agreed architecture of the project - I believe there is no need to explain why.

As we are limited with resources and are trying to fit a lot of work til the end of the year, following the process is essential for keeping things on time and producing high quality output.

File: examples/bri-3/src/bri/auth/api/auth.controller.ts
Comment (Therecanbeonlyone1969) [reply examples/bri-3/src/bri/auth/api/auth.controller.ts:19]: I agree a uuid is sufficient as a secret to be signed. The secret should be encrypted to the public key of the requesting user such that decryption requires the corresponding private key.

For replay protection, we need an incrementing nonce and the specific BPI domain to be accessed to be returned in the query response with the JWT, ideally as part of the JWT.

File: examples/bri-3/src/bri/decorators/public-endpoint.ts
Comment (Therecanbeonlyone1969) [reply examples/bri-3/src/bri/decorators/public-endpoint.ts:1]: @Manik-Jain so empty meta data signifies a login attempt?

File: examples/bri-3/src/bri/auth/api/auth.controller.ts
Comment (Therecanbeonlyone1969) [reply examples/bri-3/src/bri/auth/api/auth.controller.ts:13]: agree ... please, stick to the agreed-upon pattern @Manik-Jain, we need to be 100% constant throughout the implementation otherwise we are building in maintenance nightmares when it comes to trouble shooting, upgrades etc.

File: examples/bri-3/package.json
Comment (Therecanbeonlyone1969) [reply examples/bri-3/package.json:46]: @ognjenkurtic unless I read the docs wrong, you need local to extend the chosen strategy, namely jwt, in order to be able to do relevant configurations of the strategy through injecting a metadata object in the strategy -- which we will probably want to do soon since we need resolvable user ids. Hence, I think it is advisable to just do the extension pattern at this point. thoughts? cc @skosito @Manik-Jain @Kasshern

File: examples/bri-3/src/bri/auth/strategy/jwt.strategy.ts
Comment (Therecanbeonlyone1969) [reply examples/bri-3/src/bri/auth/strategy/jwt.strategy.ts:18]: @ognjenkurtic if the id is resolvable to its controlling keys then id is enough. Otherwise, you need to indicate which public key controls the id. I prefer the former over the latter approach.

File: examples/bri-3/src/main.ts
Comment (Therecanbeonlyone1969) [reply examples/bri-3/src/main.ts:8]: yes, super important ... please, open tracking issues to any open To-Do items. @Manik-Jain

File: examples/bri-3/src/bri/decorators/public-endpoint.ts
Comment (Manik-Jain) [reply examples/bri-3/src/bri/decorators/public-endpoint.ts:1]: @Therecanbeonlyone1969 no, any end-points annotated with this decorator will not be picked by the Nest authenticator. 

From proposal doc, this decorator covers the proposed approach
`NestJs provides a capability to either manually update each end-point to run in an authenticated manner, or we can make all the end-points authenticated by default and choose which ones can be allowed public access. I propose to use the second approach`

File: examples/bri-3/src/bri/auth/api/auth.controller.ts
Comment (Manik-Jain) [reply examples/bri-3/src/bri/auth/api/auth.controller.ts:13]: Right, this was a miss on my end. This will be pushed in the next commit

File: examples/bri-3/front-end/README.md
Comment (Manik-Jain) [reply examples/bri-3/front-end/README.md:1]: @ognjenkurtic The agreement on the call was to use a minimalistic UI to accomplish the login with MM capability. I'm not understanding what is missed here if we have a simple React application that can be extended further down the line. Could you elaborate what is the issue here, or why are we talking about Vanila JS here?

File: examples/bri-3/package.json
Comment (Manik-Jain) [reply examples/bri-3/package.json:41]: please refer to the commit here: https://github.com/eea-oasis/baseline/pull/592/commits/2367099fe113baadbdcc561f0f5359156fb74f4f

File: examples/bri-3/src/bri/identity/bpiSubjects/agents/bpiSubjectsStorage.agent.ts
Comment (Manik-Jain) [reply examples/bri-3/src/bri/identity/bpiSubjects/agents/bpiSubjectsStorage.agent.ts:69]: Do you mean BpiSubjectAgent can return anything else besides a BpiSubject? If yes, I agree to the suggested name.

File: examples/bri-3/src/main.ts
Comment (Manik-Jain) [reply examples/bri-3/src/main.ts:8]: https://github.com/eea-oasis/baseline/issues/593

**Comment by Manik-Jain** [comment]
> Some general comments:
> 
> 1. Please follow the branch naming standard of the Baseline repository.
> 2. Please provide more meaningful commit messages. General agreement is to do "This commit does..." approach where your commit message just continues the sentence. Also, commits should be much smaller.
> 3. Please follow the process we agreed on in the beginning of the project for defining the implementation direction.
>    ```
>        Create a doc with proposal -> Iterate with the team -> Write down final agreement -> Create separate smaller issues for implementation -> Implement by doing a single PR per issue
>    ```
> 
> This is important for a multitude of reasons, most notably:
> 
> * We document our design decisions in a transparent and consistent way
> * We do most of the design during the initial iterations by the team, avoiding costly rewrites and discussions once PR is created
> * We produce smaller PRs, easier to review thus with higher quality
> 
> 4. Please follow the agreed architecture of the project - I believe there is no need to explain why.
> 
> As we are limited with resources and are trying to fit a lot of work til the end of the year, following the process is essential for keeping things on time and producing high quality output.

1. Applies to others also. 
2. Can you quote the wrong commit message in this PR? 
3. **Document, iterations, agreements**: https://docs.google.com/document/d/1A4wr7Vqt9mUoDezKI5njk6eKRixGP1nUgxUe_E7pay0/edit

**Issues** : https://github.com/eea-oasis/baseline/issues/590
https://github.com/eea-oasis/baseline/issues/589
https://github.com/eea-oasis/baseline/issues/588

Since most of them need to be tested together, it was not pragmatic to segregate and test

4. I agree CQRS was missed as this whole PR was implemented in short span of time

I think we can discuss more courteously to collaborate more to SRI.

File: examples/bri-3/src/bri/auth/api/auth.controller.ts
Comment (Manik-Jain) [reply examples/bri-3/src/bri/auth/api/auth.controller.ts]: latest commit has the nonce storage changes

File: examples/bri-3/src/bri/auth/strategy/jwt.strategy.ts
Comment (Therecanbeonlyone1969) [reply examples/bri-3/src/bri/auth/strategy/jwt.strategy.ts:17]: @ognjenkurtic with all due respect, it is not, because it must be resolvable by **anyone**, and even **outside** of the BPI context since e.g. messaging between participants will be outside of the BPI context but using BpiSubject identifiers. See https://github.com/eea-oasis/baseline-standard/blob/main/core/baseline-core-v1.0-psd01.md#r41 and https://github.com/eea-oasis/baseline-standard/blob/main/core/baseline-core-v1.0-psd01.md#r42 ... we cannot assume that the BPI will be directly available for resolution of identifiers in the context of inter-party messaging.

Comment (Therecanbeonlyone1969) [reply examples/bri-3/src/bri/auth/strategy/jwt.strategy.ts:17]: @ognjenkurtic thank you for the clarification. your comment makes sense now.

**Comment by Therecanbeonlyone1969** [comment]
@Manik-Jain @ognjenkurtic @skosito @Kasshern @Ybittan what are we doing with this now?

**Comment by ognjenkurtic** [comment]
@Manik-Jain Would you like someone of us to pick up and finish this one?

**Comment by ognjenkurtic** [comment]
> @Manik-Jain @ognjenkurtic @skosito @Kasshern @Ybittan what are we doing with this now?

Either @Manik-Jain or someone from the group will finish this one.

File: examples/bri-3/front-end/README.md
Comment (skosito) [reply examples/bri-3/front-end/README.md:1]: let's just keep create-react-app, it is simple boilerplate anyways, the only thing is it should've just been separate PR so we could merge it

i understand simplicity argument, but create-react-app is simple boilerplate already, and super extensible for future... if we decided already to have UI to use for demo, we will probably have the need for couple of components down the line, let's at least make it look normal with some material UI libraries

File: examples/bri-3/front-end/src/login/Login.js
Comment (Kasshern) [inline examples/bri-3/front-end/src/login/Login.js:6]: I got this warning in the browser console : MetaMask: 'ethereum.enable()' is deprecated and may be removed in the future. Please use the 'eth_requestAccounts' RPC method instead.
For more information, see: https://eips.ethereum.org/EIPS/eip-1102

File: examples/bri-3/front-end/src/login/Login.js
Comment (Kasshern) [inline examples/bri-3/front-end/src/login/Login.js:7]: Also got this warning in the browser console: MetaMask: 'ethereum.send(...)' is deprecated and may be removed in the future. Please use 'ethereum.sendAsync(...)' or 'ethereum.request(...)' instead.
For more information, see: https://eips.ethereum.org/EIPS/eip-1193

**Review by Kasshern — CHANGES_REQUESTED** [review]
Left some comments mainly on the frontend setup seeing as some large conversations already took place or are taking place on the nestjs side. Some cleanup of react images floating around could be done but I wont die on that hill.

**Review by skosito — DISMISSED** [review]
all fixes mentioned in comments are done in this or follow up PRs, approving this one

**Review by Kasshern — DISMISSED** [review]
All concerns addressed in Clean-up PRs

**Comment by skosito** [comment]
@ognjenkurtic can you remove request for change since it is blocking the merge, if there are no more comments from your side?

**Review by skosito — APPROVED** [review]

**Review by ognjenkurtic — APPROVED** [review]

**Pull request merged**
